### PR TITLE
add mlflow handler for logging metrics only

### PR DESCRIPTION
**Pull request by thechaos16**
## Purpose
- Add support for MLflow experiment tracking during vLLM Speculator training. 
- Provides an alternative logging infrastructure for users who prefer MLflow over WandB for managing training metrics and hyperparameters.

## Description
- Implemented `MLflowHandler` for vLLM Speculator, following the design pattern of the existing `WandbHandler`.
- Enables logging of training metrics and loss values directly to an MLflow tracking server.
- Automatically initializes the MLflow run and handles structured data logging seamlessly within the speculator training loop.
- How to use (train script)
```bash
$ CUDA_VISIBLE_DEVICES="$GPUS" MLFLOW_TRACKING_URI="http://127.0.0.1:5000" torchrun \
    --standalone --nproc_per_node "$NUM_GPUS" \
    scripts/train.py \
    --verifier-name-or-path "$MODEL" \
    --data-path "$OUTPUT_DIR" \
    --hidden-states-path "$HIDDEN_STATES_DIR" \
    --save-path "$OUTPUT_DIR/checkpoints" \
    --draft-vocab-size "$DRAFT_VOCAB_SIZE" \
    --epochs "$EPOCHS" \
    --lr "$LR" \
    --total-seq-len "$SEQ_LENGTH" \
    --on-missing raise \
    --logger mlflow
```  

## Related Issue
- None

## Tests
- [Disclaimer] I used Jupyter Notebook from Vessel AI for both training and launching MLFlow, so I don't have any graphical UI for mlflow. I just tested directly from mlflow db (sqllite)
- Before running
스크린샷 2026-06-02 오전 12 03 48
- After running (experiment and metrics are filled in)  
스크린샷 2026-06-02 오전 12 05 47

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.-

### CODE REVIEW

File: src/speculators/train/logger.py [hunk 1]
```diff
@@ -479,6 +479,119 @@ def __init__(
...
+                self.client.set_experiment(self._experiment_name)
+            self._run = self.client.start_run(run_name=self.init_kwargs["name"], nested=False)
+            self._auto_end_run = True
+        else:
+            self._run = self.client.active_run()
+
+        mlflow_tags = os.getenv("MLFLOW_TAGS", None)
+        if mlflow_tags:
+            mlflow_tags = json.loads(mlflow_tags)
+            self.client.set_tags(mlflow_tags)
```
[the remark this answers is not in the document]
Comment (thechaos16) [reply src/speculators/train/logger.py:581]: I've added `import json`

File: src/speculators/train/logger.py [hunk 2]
```diff
@@ -479,6 +479,119 @@ def __init__(
...
+            self.client.log_metrics(metrics, step=step)
+    
+    def close(self):
+        """End the MLfow run and cleanup resources."""
+        if self._run is not None and self.client is not None:
+            # Only end the run if this handler started it
+            if self.client.active_run() is not None:
+                self.client.end_run()
+            self._run = None
+        super().close()
```
[the remark this answers is not in the document]
Comment (thechaos16) [reply src/speculators/train/logger.py:633]: Updated as suggested. Thanks

File: src/speculators/train/logger.py [hunk 3]
```diff
@@ -479,6 +479,119 @@ def __init__(
...
+            if self._tracking_uri:
+                self.client.set_tracking_uri(self._tracking_uri)
+            else:
+                warnings.warn(
+                    "Environment variable `MLFLOW_TRACKING_URI` is not provided and therefore will not be explicitly set." 
+                )
+        if self.client.active_run() is None or self._run_id:
+            if self._experiment_name:
+                self.client.set_experiment(self._experiment_name)
+            self._run = self.client.start_run(run_name=self.init_kwargs["name"], nested=False)
```
[the remark this answers is not in the document]
Comment (thechaos16) [reply src/speculators/train/logger.py]: I just used the same line from transformers' `MLflowCallback` (https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/integration_utils.py#L1364-L1370).

File: scripts/train.py [hunk 4]
```diff
@@ -560,7 +560,7 @@ def parse_args():
         "--logger",
         type=str,
         default="",
-        help="One of 'trackio', 'wandb', 'tensorboard' or comma separated list of them",
+        help="One of 'trackio', 'wandb', 'tensorboard', 'mlflow' or comma separated list of them",
```
Comment (fynnsu) [inline scripts/train.py:563]: Thanks for updating this, but there are also a bunch of other places in docs/readmes where we have this list. Would you be able to search the repo for `tensorboard` and add `mlflow` to the lists that come up?

File: src/speculators/train/logger.py [hunk 5]
```diff
@@ -479,6 +480,119 @@ def __init__(
...
+        self.init_kwargs = init_kwargs.copy()
+        self.init_kwargs.setdefault("dir", Path(log_dir))
+        self.init_kwargs.setdefault("name", _substitute_placeholders(run_name))
+        self.init_kwargs.setdefault("config", {})
+
+        self.client = None
+        self._package_name = "mlflow"
+        self._MAX_PARAM_VAL_LENGTH = None
+        self._MAX_PARAMS_TAGS_PER_BATCH = None
+        self._auto_end_run = False
```
Comment (fynnsu) [inline src/speculators/train/logger.py:512]: It seems like these aren't being used.

File: src/speculators/train/logger.py [hunk 6]
```diff
@@ -479,6 +480,119 @@ def __init__(
...
+        self._max_log_params = os.getenv("MLFLOW_MAX_LOG_PARAMS", 1000)
+
+        if not self.client.is_tracking_uri_set():
+            if self._tracking_uri:
+                self.client.set_tracking_uri(self._tracking_uri)
+            else:
+                warnings.warn(
+                    "Environment variable `MLFLOW_TRACKING_URI` is not provided and therefore will not be explicitly set." 
+                )
+        if self.client.active_run() is None or self._run_id:
```
Comment (fynnsu) [inline src/speculators/train/logger.py:570]: Can you explain the `active_run()` logic? Is this for resuming training? How does the user run this?

File: src/speculators/train/logger.py [hunk 7]
```diff
@@ -479,6 +480,119 @@ def __init__(
...
+        
+        flat_dict = _flatten_dict(record.msg)
+        step = getattr(record, "step", None)
+
+        # Metrics: MLflow only accepts numeric values, so split out scalars.
+        metrics = {}
+        for k, v in flat_dict.items():
+            try:
+                metrics[k] = float(v)
+            except (ValueError, TypeError):
```
Comment (fynnsu) [inline src/speculators/train/logger.py]: If you look at the tensorboard example you'll see that we handle these exception separately. A `ValueError` will be thrown by `str`s passed to `float()` whereas `TypeError` is thrown for other types. 

If `self.client.set_tag` only supports strings, then we should update the logic here to match the tensorboard handler.

File: src/speculators/train/logger.py [hunk 8]
```diff
@@ -479,6 +480,119 @@ def __init__(
...
+                # Non-numeric values (e.g. strings) aren't valid MLflow metrics.
+                self.client.set_tag(k, v)
```
Comment (fynnsu) [inline src/speculators/train/logger.py]: Another concern tho, is if this is a valid usage of tag. Above you're adding tags to the run which seem to be more higher level tags for the overall run.

Here, the `emit` function will be triggered many times, causing these tags to be repeatedly updated. Does mlflow support using tags in this way / visualizing their values over time?

Comment (fynnsu) [reply src/speculators/train/logger.py]: To be honest, I think we intentionally avoid using passing strings to the metric logger in speculators anyways so we are unlikely to hit this case, just want to understand if this will even work or if we should just ignore strings in the metric logger?

**Review by fynnsu — CHANGES_REQUESTED** [review]
Thanks for opening the pr. It's good to expand the experiment trackers we support. I haven't used mlflow before, so not super familiar with its usage / how things are visualized.

Have you been able to run this and validate the output?

File: scripts/train.py (hunk 4 above)
Comment (thechaos16) [reply scripts/train.py:563]: Thank you for your review. I updated most of documents by adding `mlflow`

File: src/speculators/train/logger.py (hunk 5 above)
Comment (thechaos16) [reply src/speculators/train/logger.py:512]: Thank you. I removed unused parameters.

File: src/speculators/train/logger.py (hunk 6 above)
Comment (thechaos16) [reply src/speculators/train/logger.py:570]: Yes, it is for resume, but there was a bug. I thought `mlflow` could resume via `run_name`, but it couldn't. So I updated a new code to get `run_id` from `run_name`. I'll attach the test result to that code.

File: src/speculators/train/logger.py (hunk 8 above)
Comment (thechaos16) [reply src/speculators/train/logger.py]: Thank you for your comment. I misunderstood mlflow's set_tag. So I copied the format from tensorboard, and if it is text format, use `log_text` instead.

mlflow will report it in `Artifact` section

스크린샷 2026-06-05 오전 1 06 29

**Comment by thechaos16** [comment]
Here's the webview of mlflow while I ran
```bash
CUDA_VISIBLE_DEVICES="$GPUS" MLFLOW_TRACKING_URI="http://127.0.0.1:5000" torchrun \
    --standalone --nproc_per_node "$NUM_GPUS" \
    scripts/train.py \
    --verifier-name-or-path "$MODEL" \
    --data-path "$OUTPUT_DIR" \
    --hidden-states-path "$HIDDEN_STATES_DIR" \
    --save-path "$OUTPUT_DIR/checkpoints" \
    --draft-vocab-size "$DRAFT_VOCAB_SIZE" \
    --epochs "$EPOCHS" \
    --lr "$LR" \
    --total-seq-len "$SEQ_LENGTH" \
    --on-missing raise \
    --logger mlflow
```

스크린샷 2026-06-05 오전 12 38 19

스크린샷 2026-06-05 오전 12 38 06

File: src/speculators/train/logger.py [hunk 9]
```diff
@@ -479,6 +480,152 @@ def __init__(
...
+            current_exp = self.client.get_experiment_by_name(self._experiment_name)
+            # find if there is same run_name in the current experiment and get its run_id to resume
+            if current_exp:
+                past_runs = self.client.search_runs(
+                    experiment_ids=[current_exp.experiment_id],
+                    filter_string=f"tags.mlflow.runName = '{self._run_name}'",
+                    max_results=1
+                )
+                if not past_runs.empty:
+                    self._run_id = past_runs.run_id[0]
```
Comment (thechaos16) [inline src/speculators/train/logger.py:559]: These lines are for getting the latest `run_id` from mlflow within the given `run_name`

It is for resume, and followings the test result.

- First run
Image

- `mlflow-test` run_name is successfully generated
Image

- Resuming run (speculators automatically detected the resumed run)
Image

- mlflow also adds newly coming metrics without adding a new run_name

Image
Image

**Comment by fynnsu** [comment]
Hey, I added a commit on top of this that simplifies some of the code / fixes a couple bugs. We'll try to land that pr (which includes your changes) instead of this one.

**Comment by thechaos16** [comment]
Thank you for letting me know.

**Comment by fynnsu** [comment]
@thechaos16 this has now landed on main.

**Comment by thechaos16** [comment]
Thank you for your hard work. This will be really helpful for our workflow.
